flink-ml-servable-core/src/main/java/org/apache/flink/ml/servable/api/DataFrame.java

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.servable.api;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.ml.servable.types.DataType;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A DataFrame consists of some number of rows, each of which has the same list of column names and
+ * data types.
+ *
+ * <p>All values in a column must have the same data type: integer, float, string etc.
+ */
+@PublicEvolving
+public class DataFrame {
+
+    private final List<String> columnNames;
+    private final List<DataType> dataTypes;
+    private final List<Row> rows;
+
+    public DataFrame(List<String> columnNames, List<DataType> dataTypes, List<Row> rows) {
+        int numColumns = columnNames.size();
+        if (dataTypes.size() != numColumns) {
+            throw new IllegalArgumentException(
+                    "The number of data types is different from the number of column names.");
+        }
+        for (Row row : rows) {
+            if (row.size() != numColumns) {
+                throw new IllegalArgumentException(
+                        "The number of values in some row is different from the number of column names");
+            }
+        }
+
+        this.columnNames = columnNames;
+        this.dataTypes = dataTypes;
+        this.rows = rows;
+    }
+
+    /** Returns a list of the names of all the columns in this DataFrame. */
+    public List<String> getColumnNames() {
+        return columnNames;
+    }
+
+    /**
+     * Returns the index of the column with the given name.
+     *
+     * @throws IllegalArgumentException if the column is not present in this table
+     */
+    public int getIndex(String name) {
+        for (int i = 0; i < columnNames.size(); i++) {
+            if (columnNames.get(i).equals(name)) {
+                return i;
+            }
+        }
+        throw new IllegalArgumentException("Failed to find the column with the given name.");
+    }
+
+    /**
+     * Returns the data type of the column with the given name.
+     *
+     * @throws IllegalArgumentException if the column is not present in this table
+     */
+    public DataType getDataType(String name) {
+        int index = getIndex(name);
+        return dataTypes.get(index);
+    }
+
+    /**
+     * Adds to this DataFrame a column with the given name, data type, and values.
+     *
+     * @throws IllegalArgumentException if the number of values is different from the number of
+     *     rows.
+     */
+    public DataFrame addColumn(String columnName, DataType dataType, List<Object> values) {
+        if (values.size() != rows.size()) {
+            throw new RuntimeException(
+                    "The number of values is different from the number of rows.");
+        }
+        columnNames.add(columnName);
+        dataTypes.add(dataType);
```

Comment:
I am not sure it is worth adding additional code to do this check. Note that there is no need to differentiate between graph build time and graph execution time since there is no Flink runtime involved in using DataFrame. 

Thus the behavior observed by users will be the same regardless of whether we do this check before add elements in this List -- users will see exception throw by flink-ml-servable classes.

I added the following statement to the Java doc of `DataFrame`. Does this address your concern?

```
 * NOTE: the columnNames and dataTypes provided to the DataFrame constructor should be immutable in order for
 * TransformerServable classes to update DataFrame with the serving results.
```

Reply:
The benefit of doing this check is that we can give users a more intuitive error message instead of `UnsupportedOperationException`, but I agree it is not necessary.

Adding docs also addresses my concern, but you need to change the `should be immutable` to `should be mutable`and also add such docs for `Row` constructor.

Reply:
Thanks for catching this. It is fixed now.